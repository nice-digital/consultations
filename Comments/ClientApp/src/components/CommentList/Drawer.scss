//@import "../../../node_modules/@nice-digital/design-system/src/stylesheets/nice-design-system";
@use "@nice-digital/nds-core/scss/spacing";
@use "@nice-digital/nds-core/scss/colours";
@use "@nice-digital/nds-core/scss/utils";
@use "@nice-digital/nds-button/scss/button";
@use "../../styles/variables" as *;
@use "../../styles/accessibility.scss" as *;

.Drawer {
  height: 100vh;
  position: fixed;
  top: 0;
  right: 0;
  z-index: 1;
  @extend %left-drop-shadow;

  // default closed state
  transform: translateX(100%);
  @extend %transition-default;
  width: calc(100% - 75px);

  &__controls {
    width: 75px;
    position: absolute;
    top: 30%;
    left: 0;
    transform: translateX(-75px);
  }

  &__control {
    font-size: 0;
    border-radius: utils.rem(10px) 0 0 utils.rem(10px);
    line-height: 1.3;
    padding: spacing.$small 0.3rem;
    text-align: left;
    border: none;
    width: 100%;
    margin-bottom: utils.rem(spacing.$x-small);
    &:focus {
      @extend .btn--cta, :focus;
      background-color: colours.$btn-cta-hover;
    }
    &:hover {
      cursor: pointer;
    }
    &--comments,
    &--questions {
      //color: $ec-white;
      box-shadow: inset -5px 0 10px rgba(0, 0, 0, 0.2);
    }
    &--comments.active,
    &--questions.active {
      background: colours.$btn-cta;
    }
    &-text {
      font-size: 0.9rem;
    }
    font-weight: bold;
  }

  &__main {
    height: 100vh;
    padding-left: utils.rem(spacing.$small);
    padding-right: utils.rem(spacing.$small);
    overflow-y: scroll;
    // overflow-scrolling: touch;
    -webkit-overflow-scrolling: touch;
    background-color: $ec-white;
    @include visibleToggleOut(
      0.4s
    ); // slightly longer than the duration of the transition
    &--open {
      @include visibleToggleIn(0.4s);
    }
  }
  &--open {
    transform: translateX(0);
  }
}

@media (min-width: 34.375em) {
  .Drawer {
    width: 430px;

    &__controls {
      width: 120px;
      transform: translateX(-120px) !important;
    }

    &__control {
      padding: utils.rem(spacing.$medium);
      font-size: 1rem;

      &-text {
        font-size: 0;
      }
    }
  }
}

@media (min-height: 12.5em) {
  .Drawer {
    &__controls {
      top: 40%;
    }
  }
}

.drawer-screenreader-button {
  @extend .screenreader-button;
  position: relative;
  &:focus,
  &:active {
    position: relative;
  }
}
