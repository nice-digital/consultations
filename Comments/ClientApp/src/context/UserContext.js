// @flow

import React from "react";
import Cookies from "js-cookie";
import { withRouter } from "react-router";

import { load } from "../data/loader";
import preload from "../data/pre-loader";

export const UserContext = React.createContext({
	updateContext: () => {},
});

type PropsType = {
	location: any,
	children: any,
	staticContext: any,
	match: any,
};

type StateType = {
	isAuthorised: boolean,
	isOrganisationCommenter: boolean,
	isLead: boolean,
	displayName: string,
	signInURL: string,
	registerURL: string,
	organisationName: string,
	initialDataLoaded: boolean,
};

export class UserProvider extends React.Component<PropsType, StateType> {
	constructor(props: PropsType) {
		super(props);

		this.state = {
			isAuthorised: false,
			isOrganisationCommenter: false,
			isLead: false,
			displayName: "",
			signInURL: "",
			registerURL: "",
			updateContext: this.updateContext,
			organisationName: null,
			initialDataLoaded: false,
		};

		const isServerSideRender = (this.props.staticContext && this.props.staticContext.preload);
		const preloadSource = isServerSideRender ? this.props.staticContext.preload.data : window.__PRELOADED__; // TODO: extract this preloaded line out to (or near) the preload endpoint method.

		const userSessionParameters = this.getUserSessionParameters(preloadSource);

		let isOrganisationCommenter = false;
		let organisationName = null;
		let isAuthorised = preloadSource ? preloadSource.isAuthorised : false;

		if (userSessionParameters.sessionCookieExistsForThisConsultation){
			const preloadedUserSessionData = preload(
				this.props.staticContext,
				"checkorganisationusersession",
				[],
				{
					consultationId: userSessionParameters.consultationId,
				},
				preloadSource,
			);

			if (preloadedUserSessionData){
				isOrganisationCommenter = preloadedUserSessionData.valid;
				isAuthorised = isAuthorised || preloadedUserSessionData.valid ; //you could be authorised by idam or by organisation session cookie.
				organisationName = preloadedUserSessionData.organisationName;
			}
		}

		if (preloadSource){
			this.state = {
				isAuthorised,
				isOrganisationCommenter: isOrganisationCommenter,
				isLead: preloadSource.isLead,
				displayName: preloadSource.displayName,
				signInURL: preloadSource.signInURL,
				registerURL: preloadSource.registerURL,
				updateContext: this.updateContext,
				organisationName: organisationName,
				initialDataLoaded: true,
			};
			if (this.props.staticContext) {
				this.props.staticContext.analyticsGlobals.isSignedIn = preloadSource.isAuthorised;
			}
		}
	}

	setStateForValidSessionCookie = (organisationName) => {
		this.setState({
			isAuthorised: true,
			isOrganisationCommenter: true,
			organisationName,
		});
	}

	loadUser = (returnURL) => {
		load("user", undefined, [], { returnURL, cachebust: new Date().getTime() })
			.then(
				res => {
					const signInURL = res.data.signInURL;
					this.setState({
						isAuthorised: res.data.isAuthenticatedByAccounts,
						displayName: res.data.displayName,
						signInURL: signInURL,
						registerURL: res.data.registerURL,
						isLead: (res.data.organisationsAssignedAsLead && res.data.organisationsAssignedAsLead.length > 0),
					});
					//update signin links in global nav here. because SSR isn't rendering them right on the server.
					var signInLinks = document.getElementById("global-nav-header").querySelectorAll("a[href*='account/login']");
					for (var i=0; i < signInLinks.length; i++) {
						signInLinks[i].setAttribute("href", signInURL);
					}
				},
			)
			.then(
				this.checkSessionId()
					.then(data => {
						if (data.validityAndOrganisationName?.valid === true) {
							this.setStateForValidSessionCookie(data.validityAndOrganisationName.organisationName);
						}
					}),
			);
	}

	getFeatureFlags = async () => {
		load("getfeatureflags", undefined, [], {})
			.then(response =>
				this.setState({
					organisationalCommentingFeature: response.data.OrganisationalCommenting,
				}))
			.catch(err => {
				console.log(JSON.stringify(err));
			});
	}

	//unfortunately the context is above the routes, so this.props.match is always null, so we can't pull the consultation id out of there. hence we're falling back to regex.
	getConsultationId = () =>{
		const regex = /^\/(\d+)\/(review|submitted|\d+\/[a-z-A-Z0-9]+)/;
		const pathname = this.props.location.pathname;

		if (!pathname)
			return;

		const matches = regex.exec(pathname);
		if (!matches || matches.length !== 3)
			return;

		return matches[1];
	}

	checkSessionId = async () => {
		const userSessionParameters = this.getUserSessionParameters(window.__PRELOADED__);

		if (!userSessionParameters.sessionCookieExistsForThisConsultation)
			return await {
				validityAndOrganisationName: {valid: false},
				userSessionParameters: userSessionParameters,
			};

		const validityAndOrganisationName = load(
			"checkorganisationusersession",
			undefined,
			[],
			{
				consultationId: userSessionParameters.consultationId,
			})
			.then(response => response.data)
			.catch(err => {
				console.log(JSON.stringify(err));
				return {valid: false};
			});
		return {
			validityAndOrganisationName: await validityAndOrganisationName,
			userSessionParameters: userSessionParameters,
		};
	}

	getUserSessionParameters = (preloadSource) => {
		const consultationId = this.getConsultationId();
		const cookieName = `ConsultationSession-${consultationId}`;
		const cookieString = preloadSource?.cookies || "";
		const sessionCookieExistsForThisConsultation =  (Cookies.get(cookieName) || cookieString.indexOf(cookieName) !== -1);
		return {consultationId, sessionCookieExistsForThisConsultation};
	}

	//a child component can call this method to update the context in case a session cookie has been set.
	updateContext = () => {
		this.checkSessionId()
			.then(data => {
				if (data.validityAndOrganisationName.valid === true) {
					this.setStateForValidSessionCookie(data.validityAndOrganisationName.organisationName);
				}
			});
	}

	loadContext = () => {
		this.loadUser(this.props.location.pathname);
	}

	componentDidMount() {
		this.loadContext(); //this is needed here as the sign in url isn't right on SSR. TODO: fix SSR.
	}

	componentDidUpdate(prevProps) {
		if (this.props.location !== prevProps.location) { //fired when the route changes, as you might be authenticated for some routes, but not others (e.g. org commenting cookie)
			this.loadContext();
		}
	}

	render() {
		return (
			<UserContext.Provider value={this.state}>
				{this.props.children}
			</UserContext.Provider>
		);
	}
}

export default withRouter(UserProvider);
