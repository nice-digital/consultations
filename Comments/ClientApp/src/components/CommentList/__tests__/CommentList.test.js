/* global jest */

import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { CommentList } from "../CommentList";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { LiveAnnouncer } from "react-aria-live";
import toJson from "enzyme-to-json";

import { generateUrl } from "../../../data/loader";
import sampleComments from "./sampleComments.json";
import sampleQuestions from "./sampleQuestions.json";
import sampleQuestionsWithNoAnswers from "./sampleQuestionsWithNoAnswers.json";
import sampleConsultation from "./sampleconsultation.json";
import emptyCommentsResponse from "./EmptyCommentsResponse.json";
import { nextTick } from "../../../helpers/utils";

import { createQuestionPdf } from "../../QuestionView/QuestionViewDocument";
import { UserContext } from "../../../context/UserContext";

const mock = new MockAdapter(axios);

jest.mock("../../QuestionView/QuestionViewDocument");

describe("[ClientApp] ", () => {
	describe("CommentList Component", () => {
		const fakeProps = {
			match: {
				url: "/1/1/introduction",
				params: {
					consultationId: 1,
					documentId: 1,
					chapterSlug: "introduction",
				},
			},
			location: {
				pathname: "",
			},
			comment: {
				commentId: 1,
			},
		};

		let contextWrapper = null;

		beforeEach(() => {
			contextWrapper = {
				wrappingComponent: UserContext.Provider,
				wrappingComponentProps: {
					value: { isAuthorised: true, isOrganisationCommenter: false, isLead: false },
				},
			};
		});

		afterEach(() => {
			mock.reset();
		});

		describe("Handlers", () => {
			it("save handler put's to the api with updated comment", async done => {
				const commentToUpdate = sampleComments.comments[0];

				mock.onGet(generateUrl("comments", undefined, [], { sourceURI: fakeProps.match.url }))
					.reply(200, sampleComments);

				mock.onPut("/consultations/api/Comment/" + commentToUpdate.commentId)
					.reply(config => {
						expect(JSON.parse(config.data)).toEqual(commentToUpdate);
						done();
						return [200, commentToUpdate];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);
				wrapper.instance().saveCommentHandler(new Event("click"), commentToUpdate);
			});

			it("save handler updates the correct item in the comments array once the api has returned new data", async () => {
				const commentToUpdate = sampleComments.comments[1];

				mock.onGet()
					.reply(200, sampleComments);

				mock.onPut("/consultations/api/Comment/" + commentToUpdate.commentId)
					.reply(() => {
						const sampleCommentsUpdated = {...sampleComments.comments[1]};
						sampleCommentsUpdated.commentText = "New updated text";
						return [200, sampleCommentsUpdated];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);
				wrapper.instance().saveCommentHandler(new Event("click"), commentToUpdate);

				await nextTick();
				wrapper.update();

				expect(wrapper.state().comments[1].commentText).toEqual("New updated text");
			});

			it("new comment should add an entry in the array with negative id", () => {
				mock.onGet()
					.reply(() => {
						return [200, { comments: [] }];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);

				expect(wrapper.state().comments.length).toEqual(0);

				wrapper.instance().newComment(null, {order: "0"});

				expect(wrapper.state().comments.length).toEqual(1);
				expect(wrapper.state().comments[0].commentId).toEqual(-1);
			});

			it("2 new comments should decrement the negative commentId without conflicting", () => {
				mock.onGet()
					.reply(() => {
						return [200, { comments: [] }];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);

				expect(wrapper.state().comments.length).toEqual(0);

				wrapper.instance().newComment(null, {order: "0"});
				wrapper.instance().newComment(null, {order: "0"});

				expect(wrapper.state().comments.length).toEqual(2);
				expect(wrapper.state().comments[0].commentId).toEqual(-2);
				expect(wrapper.state().comments[1].commentId).toEqual(-1);
			});

			it("where there are existing comments in the array, inserting 2 new comments should still decrement the negative commentId without conflicting", () => {
				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(200, sampleComments);

				const wrapper = shallow(<CommentList {...fakeProps} />);

				expect(wrapper.state().comments.length).toEqual(0);

				wrapper.instance().newComment(null, {order: "0"});
				wrapper.instance().newComment(null, {order: "0"});

				expect(wrapper.state().comments.length).toEqual(2);
				expect(wrapper.state().comments[0].commentId).toEqual(-2);
				expect(wrapper.state().comments[1].commentId).toEqual(-1);
			});

			it("save handler posts to the api with new comment with correct path from generateUrl", async done => {
				const commentToInsert = {
					commentId: -1,
					commentText: "a newly created comment",
				};

				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(200, emptyCommentsResponse);

				mock.onPost(generateUrl("newcomment"))
					.reply(config => {
						expect(JSON.parse(config.data)).toEqual(commentToInsert);
						expect(config.url).toEqual("/consultations/api/Comment");
						done();
						return [200, commentToInsert];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);
				wrapper.instance().saveCommentHandler(new Event("click"), commentToInsert);
			});

			it("delete handler called with negative number removes item from array and calls aria announcer with correct message", async () => {
				mock.onGet()
					.reply(200, sampleComments);

				fakeProps.announceAssertive = jest.fn();

				const wrapper = shallow(<CommentList {...fakeProps} />);

				await nextTick();
				wrapper.update();

				expect(wrapper.state().comments.length).toEqual(5);

				wrapper.instance().newComment(null, {
					sourceURI: "/1/1/introduction",
					commentText: "",
					order: "0",
				});

				expect(wrapper.state().comments.length).toEqual(6);

				wrapper.instance().deleteCommentHandler(new Event("click"), -1);

				expect(wrapper.state().comments.length).toEqual(5);
				expect(fakeProps.announceAssertive).toHaveBeenCalledWith("Comment deleted", expect.any(String));
			});

			it("delete handler called with positive number hits the correct delete endpoint", async () => {
				mock.onGet()
					.reply(200, sampleComments);

				mock.onDelete(generateUrl("editcomment", undefined, [1004]))
					.reply(config => {
						expect(config.url).toEqual("/consultations/api/Comment/1004");
						return [200, {}];
					});

				fakeProps.announceAssertive = jest.fn();

				const wrapper = shallow(<CommentList {...fakeProps} />);

				await nextTick();
				wrapper.update();

				expect(wrapper.state().comments.length).toEqual(5);

				wrapper.instance().deleteCommentHandler(new Event("click"), 1004);

				await nextTick();
				wrapper.update();

				expect(wrapper.state().comments.length).toEqual(4);
			});
		});

		describe("Authorisation", () => {
			it("should show login panel and not render any comment boxes without authorisation", async () => {
				mock.onGet()
					.reply(200, sampleComments);

				contextWrapper.wrappingComponentProps.value.isAuthorised = false;

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} />
						</LiveAnnouncer>
					</MemoryRouter>,
					contextWrapper,
				);

				await nextTick();
				wrapper.update();

				expect(wrapper.find("LoginPanel").length).toEqual(1);
				expect(wrapper.find("CommentBox").length).toEqual(0);
			});

			it("should show 5 (current user) comment boxes with authorisation and sample data", async () => {
				mock.onGet()
					.reply(200, sampleComments);

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} />
						</LiveAnnouncer>
					</MemoryRouter>,
					contextWrapper,
				);

				await nextTick();
				wrapper.update();

				expect(wrapper.find("textarea").length).toEqual(5);
			});

			it("should match snapshot post login, with 5 personal comments and 5 read only comments from others", async () => {
				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(200, sampleComments);

				mock.onGet(generateUrl("commentsForOtherOrgCommenters", undefined, [], { sourceURI: fakeProps.match.url }))
					.reply(200, sampleComments);

				contextWrapper.wrappingComponentProps.value.isOrganisationCommenter = true;

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} />
						</LiveAnnouncer>
					</MemoryRouter>,
					contextWrapper,
				);

				await nextTick();
				wrapper.update();

				expect(
					toJson(wrapper, {
						noKey: true,
						mode: "deep",
					}),
				).toMatchSnapshot();
			});
		});

		describe("Organisational Commenting", () => {
			it("should hit the commentsForOtherOrgCommenters endpoint (when using the code) and populate comment list with read only comments", async () => {
				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(200, emptyCommentsResponse);

				mock.onGet(generateUrl("commentsForOtherOrgCommenters", undefined, [], { sourceURI: fakeProps.match.url }))
					.reply(200, sampleComments);

				contextWrapper.wrappingComponentProps.value.isOrganisationCommenter = true;

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} />
						</LiveAnnouncer>
					</MemoryRouter>,
					contextWrapper,
				);

				await nextTick();
				wrapper.update();

				expect(wrapper.find("textarea").filter({ disabled: false }).length).toEqual(0);
				expect(wrapper.find("textarea").filter({ disabled: true }).length).toEqual(5);
				expect(wrapper.find("button").filter({ children: "Delete" }).length).toEqual(0);
			});

			it("should hit the commentsForOtherOrgCommenters endpoint (when org commenter) and populate question list with read only comments", async () => {
				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(200, sampleQuestionsWithNoAnswers);

				mock.onGet(generateUrl("commentsForOtherOrgCommenters", undefined, [], { sourceURI: fakeProps.match.url }))
					.reply(200, sampleQuestions);

				contextWrapper.wrappingComponentProps.value.isOrganisationCommenter = true;

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} />
						</LiveAnnouncer>
					</MemoryRouter>,
					contextWrapper,
				);

				await nextTick();
				wrapper.update();

				// new answer boxes appear as the user hasn't added any themselves
				expect(wrapper.find("textarea").filter({ disabled: false }).length).toEqual(2);

				// readonly / disabled answer boxe for three existing questions
				expect(wrapper.find("textarea").filter({ disabled: true }).length).toEqual(3);
			});
		});

		describe("API", () => {
			it("should make get api call for comments with correct path and update state with empty array", async done => {
				mock.onGet(generateUrl("comments", undefined, [], {	sourceURI: fakeProps.match.url }))
					.reply(config => {
						expect(config.url).toEqual(
							"/consultations/api/Comments?sourceURI=%2F1%2F1%2Fintroduction",
						);
						done();
						return [200, emptyCommentsResponse];
					});

				const wrapper = shallow(<CommentList {...fakeProps} />);

				expect(Array.isArray(wrapper.state().comments)).toEqual(true);
				expect(wrapper.state().comments.length).toEqual(0);
			});
		});

		describe("Download responses", () => {
			it("should call createQuestionPDF with title, end date, and questions when the download questions button is clicked", async () => {
				mock.onGet("/consultations/api/Comments?sourceURI=%2F1%2F1%2Fintroduction")
					.reply(200, sampleComments)
					.onGet("/consultations/api/Consultation?consultationId=1&isReview=false")
					.reply(200, sampleConsultation);

				const questionsForPDF = sampleComments.questions;
				const titleForPDF = sampleConsultation.title;
				const endDate = sampleComments.consultationState.endDate;
				const getTitleFunction = () => { return titleForPDF;};

				const wrapper = mount(
					<MemoryRouter>
						<LiveAnnouncer>
							<CommentList {...fakeProps} getTitleFunction={getTitleFunction} />
						</LiveAnnouncer>
					</MemoryRouter>,
				);

				await nextTick();
				wrapper.update();

				const button = wrapper.find("#js-create-question-pdf");

				button.simulate("click");

				expect(createQuestionPdf).toHaveBeenCalledTimes(1);
				expect(createQuestionPdf).toHaveBeenCalledWith(questionsForPDF, titleForPDF, endDate);
			});
		});
	});
});
