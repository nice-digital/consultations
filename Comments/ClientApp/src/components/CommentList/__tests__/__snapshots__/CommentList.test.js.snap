// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot post login, with 5 personal comments and 5 read only comments from others 1`] = `
<div>
  <section
    aria-label="Commenting panel"
    class="Drawer "
  >
    <div
      class="Drawer__controls"
    >
      <button
        aria-controls="comments-panel"
        aria-haspopup="true"
        aria-label="Open the commenting panel"
        class="Drawer__control Drawer__control--comments active"
        data-qa-sel="open-commenting-panel"
        id="js-drawer-toggleopen-comments"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="Drawer__control-text"
          data-qa-sel="close-commenting-panel"
        >
          Comments
        </span>
        Open comments
      </button>
      <button
        aria-controls="questions-panel"
        aria-haspopup="true"
        aria-label="Open the questions panel"
        class="Drawer__control Drawer__control--questions active"
        data-qa-sel="open-questions-panel"
        id="js-drawer-toggleopen-questions"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="Drawer__control-text"
          data-qa-sel="close-questions-panel"
        >
          Questions
        </span>
        Open questions
      </button>
    </div>
    <div
      aria-disabled="true"
      class="Drawer__main Drawer__main--closed"
      data-qa-sel="comment-panel"
      id="comments-panel"
    >
      <div
        data-qa-sel="comment-list-wrapper"
      >
        <a
          class="btn btn--cta mt--c"
          data-qa-sel="review-all-comments"
          href="/1/review"
        >
          Review your response before submitting   
          <span
            aria-hidden="true"
            class="icon icon--chevron-right"
          />
        </a>
        <div
          class="show"
        >
          <ul
            class="CommentList list--unstyled mt--0"
          >
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1002"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      id="Comment1002"
                      tabindex="0"
                    >
                      first comment
                    </textarea>
                  </div>
                  <span
                    class="ml--0 mb--0 CommentBox__savedIndicator"
                  >
                    Saved
                  </span>
                  <button
                    class="btn mr--0 mb--0 right"
                    data-qa-sel="delete-comment-button"
                  >
                    Delete
                  </button>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1003"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      id="Comment1003"
                      tabindex="0"
                    >
                      second comment
                    </textarea>
                  </div>
                  <span
                    class="ml--0 mb--0 CommentBox__savedIndicator"
                  >
                    Saved
                  </span>
                  <button
                    class="btn mr--0 mb--0 right"
                    data-qa-sel="delete-comment-button"
                  >
                    Delete
                  </button>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1004"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      id="Comment1004"
                      tabindex="0"
                    >
                      third comment
                    </textarea>
                  </div>
                  <span
                    class="ml--0 mb--0 CommentBox__savedIndicator"
                  >
                    Saved
                  </span>
                  <button
                    class="btn mr--0 mb--0 right"
                    data-qa-sel="delete-comment-button"
                  >
                    Delete
                  </button>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1005"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      id="Comment1005"
                      tabindex="0"
                    >
                      fourth comment
                    </textarea>
                  </div>
                  <span
                    class="ml--0 mb--0 CommentBox__savedIndicator"
                  >
                    Saved
                  </span>
                  <button
                    class="btn mr--0 mb--0 right"
                    data-qa-sel="delete-comment-button"
                  >
                    Delete
                  </button>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1006"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      id="Comment1006"
                      tabindex="0"
                    >
                      fifth comment
                    </textarea>
                  </div>
                  <span
                    class="ml--0 mb--0 CommentBox__savedIndicator"
                  >
                    Saved
                  </span>
                  <button
                    class="btn mr--0 mb--0 right"
                    data-qa-sel="delete-comment-button"
                  >
                    Delete
                  </button>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1002"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      disabled=""
                      id="Comment1002"
                      tabindex="0"
                    >
                      first comment
                    </textarea>
                  </div>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1003"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      disabled=""
                      id="Comment1003"
                      tabindex="0"
                    >
                      second comment
                    </textarea>
                  </div>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1004"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      disabled=""
                      id="Comment1004"
                      tabindex="0"
                    >
                      third comment
                    </textarea>
                  </div>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1005"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      disabled=""
                      id="Comment1005"
                      tabindex="0"
                    >
                      fourth comment
                    </textarea>
                  </div>
                </form>
              </section>
            </li>
            <li
              class="CommentBox"
            >
              <section
                role="form"
              >
                <h3
                  class="CommentBox__title mt--0 mb--d"
                  data-qa-sel="comment-box-title"
                >
                  Comment on 
                  <span
                    class="text-lowercase"
                  />
                </h3>
                <form
                  class="mb--0"
                >
                  <div
                    class="form__group form__group--textarea mb--b"
                  >
                    <label
                      class="form__label visually-hidden"
                      for="Comment1006"
                    >
                      Comment on 
                      , 
                    </label>
                    <textarea
                      class="form__input form__input--textarea"
                      data-hj-whitelist="true"
                      data-qa-sel="Comment-text-area"
                      disabled=""
                      id="Comment1006"
                      tabindex="0"
                    >
                      fifth comment
                    </textarea>
                  </div>
                </form>
              </section>
            </li>
          </ul>
          <button
            class="drawer-screenreader-button"
          >
            Close the commenting panel
          </button>
        </div>
        <div
          class="hide"
        >
          <button
            class="btn btn--primary"
            data-qa-sel="create-question-pdf"
            id="js-create-question-pdf"
          >
            Download questions (PDF)
          </button>
          <p
            class="mt--0"
          >
            Please answer the following questions
          </p>
          <ul
            class="CommentList list--unstyled mt--0"
          />
          <button
            class="drawer-screenreader-button"
          >
            Close the questions panel
          </button>
        </div>
      </div>
    </div>
  </section>
  <div>
    <div
      aria-live="assertive"
      role="log"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; white-space: nowrap; padding: 0px; width: 1px; position: absolute;"
    >
      Comments and questions panel closed
    </div>
    <div
      aria-live="assertive"
      role="log"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; white-space: nowrap; padding: 0px; width: 1px; position: absolute;"
    />
    <div
      aria-live="polite"
      role="log"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; white-space: nowrap; padding: 0px; width: 1px; position: absolute;"
    />
    <div
      aria-live="polite"
      role="log"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; white-space: nowrap; padding: 0px; width: 1px; position: absolute;"
    />
  </div>
</div>
`;
